### DISCUSSION TITLE
Alternative way to activate Menu (Super+RMB)

### DISCUSSION THREAD

**Discussion by Leleat** [Schneegans/Fly-Pie#148] [Ideas & Feature Discussions]
Currently you can only activate Fly Pie with a keyboard shortcut. It would be pretty cool, if there was a way to activate Fly Pie with just `Super` + `RMB`. That way it would feel more integrated with GNOME since `Super` is used as the main key all over the place. It would also add another activation method for an item. Instead of having to open the Pie with a shortcut, then dragging the mouse to an item and then clicking the item / waiting for a timeout, you could just directly open the Pie with `Super` + `RMB` (on start of the press) and activate an item as soon as `RMB` is released.

If you want to try out how this feels, I had a very basic pieMenu in my own extension (https://extensions.gnome.org/extension/3733/tiling-assistant/?c=114572). But I removed it in the latest version in favor of Fly Pie. So if you want to try it out, you need to install v23. Note that I only replaced the `Super` + `RMB` on window since my pie menu was only used for window management. I didn't hooking into the Super + RMB on the Desktop.

If you do want to implement this, these 2 files / functions seem to be relevant:

- main.wm._windowMenuManager.showWindowMenuForWindow (I used this one)
- backgroundMenu.js for Super + RMB on the Desktop

**Comment by Schneegans** [comment]
I would be extremely happy if it would be possible to bind menus to mouse buttons. In fact, this was a key feature of Fly-Pie's predecessor Gnome-Pie. With Gnome-Pie it was not only possible to use <kbd>Any Modifier</kbd>+<kbd>Any Mouse Button</kbd> but also to bind a single mouse button. This is useful if you have one of these mice with plenty of buttons. Furthermore, there was a _delayed_ binding which would show a menu if the mouse button was held down for a couple of milliseconds. This way you could also bind pie menus directly to your <kbd>RMB</kbd> (or even the <kbd>LMB</kbd> but then it becomes difficult to drag things around :smile:).

But sadly none of this stuff works (for good reason) on Wayland. This was only possible to all the security flaws of X11.

The main reason to make Fly-Pie a GNOME Shell extension (and not a standalone application) was to circumvent some of the Wayland-related restrictions for "normal" client applications. Yet I haven't found a way to properly intercept mouse events to do things like described above. However, to be honest, I haven't search thoroughly as I was pretty sure that these global key bindings are managed by Mutter and all other input is directly sent to the client windows with no way to intercept it from an extension (would be kind of flawed security-wise if as possible...).

Nevertheless, what you're suggesting looks somewhat promising. Was it possible to open the pie menu in your extension from any position on the window or only when clicking the window decorations?

**Comment by Schneegans** [comment]
There's also the <kbd>Super</kbd>+<kbd>LMB</kbd> which somehow gets intercepted by GNOME Shell to start window dragging. Would be great to find this in the source...

**Comment by Leleat** [comment]
> Was it possible to open the pie menu in your extension from any position on the window or only when clicking the window decorations?

Yes, here is a short demo gif:

![output](https://user-images.githubusercontent.com/53054301/137537163-3e1b6c6e-8caa-41d7-898f-4d83f3f8c613.gif)

> There's also the Super+LMB which somehow gets intercepted by GNOME Shell to start window dragging. Would be great to find this in the source...

I am not very familiar with window management and the like so take everything I say from now on with a big grain of salt...

But I don't think we'd be able to override the `Super + LMB / MMB` behaviour. That's probably not passed on to GNOME Shell since there is no need to. `Super + RMB` is 'hardcoded' as a signal and passed to gnome shell [here](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/main/js/ui/windowManager.js#L588) so that GNOME Shell can open the window menu when necessary. This is the window menu btw

![scr](https://user-images.githubusercontent.com/53054301/137538248-9c74b8f0-e564-464b-bd2d-a79404730925.png)

The window menu opens / the 'show-window-menu' signal gets fired when you `Super + RMB` a window or `RMB` the titlebar. The differentiation between the 2 cases is no problem since you can just check for the `Super` key. So you could replace the `Super+RMB` case by overriding `main.wm._windowMenuManager.showWindowMenuForWindow`.

**Comment by Leleat** [comment]
> main.wm._windowMenuManager.showWindowMenuForWindow

This ofc only works for `super+rmb`-ing a window. You' also need to cover the case when `super+rmb`-ing the desktop, which can probably be done in [backgroundMenu.js](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/main/js/ui/backgroundMenu.js). Although I guess, this is aa kinda ugly solution...

![scr](https://user-images.githubusercontent.com/53054301/137539616-9862fb2e-7d5a-4773-9a52-56d2429f0934.png)

**Comment by Schneegans** [comment]
Ahh, I attached a mouse to my laptop and now <kbd>Super</kbd>+<kbd>RMB</kbd> works as you describe. For some reason I cannot summon the menu with the touchpad...

There are also some other places where this wouldn't work such as the overview, the panel, or maybe the dash if you have it visible by using some extension. But anyway, you're right, this would feel well integrated. So it would be possible to replace this context menu with **one** of the configured pie menus.

Btw, there is [global.get_pointer()](https://gjs-docs.gnome.org/shell01~0.1_api/shell.global#method-get_pointer) which gives you pointer position and modifiers (including pressed buttons). But we would have to poll this every few milliseconds and even the we would miss short clicks, I guess. But this is _nearly_ the required API...

**Comment by Leleat** [comment]
>  one of the configured pie menus.

Yep, that was what I meant. An alternative /additional activation method instead / besides of a keyboard shortcut.

> Ahh, I attached a mouse to my laptop and now Super+RMB works as you describe. For some reason I cannot summon the menu with the touchpad...

Do you mean GNOME's window menu or the pie menu from Tiling Assistant? I don't have a problem with opening the window menu on my notebook. If it's the later, then it's intentional (I guess). You open my pie Menu with RMB and activate an item with the release of that very same click (You don't need to hold Super though). On a laptop you can't hold the RMB, so you would immediately leave my pieMenu and not activate an item since you are still within the deadzone. (And you won't see it because I've delayed the visual appearance of the pie menu, like I requested in in the other discussion :) ).

> Btw, there is global.get_pointer() which gives you pointer position and modifiers (including pressed buttons). But we would have to poll this every few milliseconds and even the we would miss short clicks, I guess. But this is nearly the required API...

Sorry, I don't really follow here. Why would you need to poll every millisecond? My idea is: You would just open the Fly Pie with the signals. The user doesn't need to hold `Super`, they only need to hold the `RMB`. If they let go of `RMB, an item gets activated.

**Comment by Schneegans** [comment]
I meant GNOME's window menu. I use two-finger-tap to open the context menu usually and this does not open GNOME's window menu when combined with <kbd>Super</kbd>.

> Sorry, I don't really follow here. Why would you need to poll every millisecond? My idea is: You would just open the Fly Pie with the signals. The user doesn't need to hold Super, they only need to hold the RMB. If they let go of `RMB, an item gets activated.

Sorry for the confusion: I was thinking about the possibility of a more general solution allowing to bind any mouse button with any modifier. This _could_ nearly work with `global.get_pointer()`. But it wouldn't be reliable as we would miss short clicks.

**Comment by Leleat** [comment]
> Sorry for the confusion: I was thinking about the possibility of a more general solution allowing to bind any mouse button with any modifier. This could nearly work with global.get_pointer(). But it wouldn't be reliable as we would miss short clicks.

Ah, I see what you mean now.

> I meant GNOME's window menu. I use two-finger-tap to open the context menu usually and this does not open GNOME's window menu when combined with Super

It works for me with super + 2 finger tap; maybe a bug 🤷

**Comment by Schneegans** [comment]
I just did a quick test. Polling every 20 milliseconds is sufficient for me to capture even the shortest clicks I am able to perform :rocket: :computer_mouse:  

```javascript
GLib.timeout_add(GLib.PRIORITY_DEFAULT, 20, () => {
    const [x, y, mods] = global.get_pointer();

    if (mods & Clutter.ModifierType.BUTTON3_MASK &&
        mods & Clutter.ModifierType.MOD4_MASK) {
      log('click!');
    }

    return true;
});
```

This is kind of hacky, but it would greatly improve the possibilities and usability of Fly-Pie! Performance-wise it should be neglectable, every 20 milliseconds isn't too much...

**Comment by Leleat** [comment]
> I just did a quick test. Polling every 20 milliseconds is sufficient for me to capture even the shortest clicks I am able to perform rocket computer_mouse

Cool, any plans when / where (in the settings) this will land in the `develop` branch?

**Comment by Schneegans** [comment]
I guess it will take some time. Fly-Pie 8 will be released quite soonishly with a focus on GNOME 3.3x compatibility. Thereafter I will look into this one and your other feature request. But both will need quite some work.

This one here will be tricky as Fly-Pie is currently using the "normal" GTK accelerator functions (e.g. `gtk_accelerator_get_label`, `gtk_accelerator_parse`, etc.). These do not know anything about mouse buttons, so I will have to partially re-implement them in gjs. Furthermore I expect some pitfalls: For example, I will not be able to prevent the mouse click from being sent to the currently focused window (maybe the button-up event will be captured by Fly-Pie). This will be an issue for most Modifier+Mouse Button combinations...

**Comment by Schneegans** [comment]
I'll create an issue anyways and report there any findings. If you find another approach for binding mouse buttons, I would be very happy as this seems to be quite ugly...

**Reply by Schneegans** [reply]
Here we go: #149

**Comment by Schneegans** [comment]
The plan to bind any mouse button + modifier does not succeed, as I have no way to prevent the original event. Basically, I _can_ notice when the user issues <kbd>Ctrl</kbd>+<kbd>RMB</kbd>, but I _cannot_ prevent the context menu from showing up.

So I decided to attempt the approach proposed by you in the first place. I intercept <kbd>Super</kbd>+<kbd>RMB</kbd> on a window like you did and in all other places by directly intercepting the corresponding events on the `global.stage`. Therefore it works anywhere!

I submitted an initial implementation of this feature to the `feature/super-rmb` branch (#169). You can test it by [installing the current development version of Fly-Pie](https://github.com/Schneegans/Fly-Pie/blob/develop/docs/installation.md#c-cloning-the-latest-version-with-git) and checking out this branch! (Feel free to ask is anything is unclear in this guide). Then:
* Open the menu editor of Fly-Pie.
* Select any of your menus.
* Check the "Bind to Super + RMB" option.

What do you think?

In the coming days I check whether <kbd>Super</kbd>+<kbd>LMB</kbd> could be possible as well....

**Reply by Leleat** [reply]
Thanks! I tested it out for a few minutes and noticed the following 2 things:

- the `Super`+`Whatever` shortcuts don't seem to work anymore
- releasing the initial `RMB` in the center doesn't abort the pie menu and a quick second `RMB` doesn't get registered. So it seems like there is no quick way to abort/escape the pie menu. This feels especially slow, if you use the `Display Timeout` setting since you need to 'wait' for the pie menu to escape it.

PS.

I don't know, if you have a specific reason for having the version nr in `metadata.json` be higher than the EGO version. If there is no specific reason, I'd suggest to keep the development version nr the same as the EGO version. This is helpful for user, who want to test specific git versions *at a certain point in time* but still want to stick to EGO to get the automatic updates (like me atm 😉). Right now, after testing specific features, I need to downgrade again to get the actual next full release from EGO. If the development version was the same, I could just stick to the current development version and get the full version update from EGO.

**Reply by Leleat** [reply]
I've got a freeze/crash. Don't know, if it's related to Fly-Pie but there were some errors in the journalctl **around** that time (Didn't check the exact time of the freeze/crash). There were also errors from `compiz-alike-windows-effect`... so maybe not related to Fly-Pie. If you need more information to debug this, please do tell me what you need since I have no clue what would be helpful in this case.

PS. I can no longer reproduce the crash.

```
Nov 21 22:57:15 fedx gnome-shell[1545]: [flypie@schneegans.github.com/src/extension/Daemon.js:101:15] Failed to open a Fly-Pie menu: An unknown error occurred.
Nov 21 22:57:15 fedx gnome-shell[1545]: [flypie@schneegans.github.com/src/extension/Daemon.js:377:13] Failed to show menu: TypeError: item is null
Nov 21 22:57:11 fedx gnome-shell[1545]: [flypie@schneegans.github.com/src/extension/Daemon.js:101:15] Failed to open a Fly-Pie menu: An unknown error occurred.
Nov 21 22:57:11 fedx gnome-shell[1545]: [flypie@schneegans.github.com/src/extension/Daemon.js:377:13] Failed to show menu: TypeError: item is null
Nov 21 22:57:10 fedx gnome-shell[1545]: [flypie@schneegans.github.com/src/extension/Daemon.js:101:15] Failed to open a Fly-Pie menu: An unknown error occurred.
Nov 21 22:57:10 fedx gnome-shell[1545]: [flypie@schneegans.github.com/src/extension/Daemon.js:377:13] Failed to show menu: TypeError: item is null
Nov 21 22:57:10 fedx gnome-shell[1545]: JS ERROR: TypeError: child is undefined
                                        Menu/<@/home/timmy/.local/share/gnome-shell/extensions/flypie@schneegans.github.com/src/extension/Menu.js:390:26
                                        emitSelection@/home/timmy/.local/share/gnome-shell/extensions/flypie@schneegans.github.com/src/extension/SelectionWedges.js:353:14
                                        onMotionEvent/this._stroke.pauseTimeout<@/home/timmy/.local/share/gnome-shell/extensions/flypie@schneegans.github.com/src/extension/SelectionWedges.js:496:24
Nov 21 22:57:10 fedx gnome-shell[1545]: #3   556f89888818 i   /home/timmy/.local/share/gnome-shell/extensions/flypie@schneegans.github.com/src/extension/SelectionWedges.js:496 (3cb0807d01f0 @ 13)
Nov 21 22:57:10 fedx gnome-shell[1545]: #2   7ffe3a081fb0 b   /home/timmy/.local/share/gnome-shell/extensions/flypie@schneegans.github.com/src/extension/SelectionWedges.js:353 (3cb0807d0290 @ 173)
Nov 21 22:57:10 fedx gnome-shell[1545]: #1   556f898888a8 i   /home/timmy/.local/share/gnome-shell/extensions/flypie@schneegans.github.com/src/extension/Menu.js:384 (3cb0807ce510 @ 75)
Nov 21 22:57:10 fedx gnome-shell[1545]: #0   7ffe3a080df0 b   /home/timmy/.local/share/gnome-shell/extensions/flypie@schneegans.github.com/src/extension/MenuItem.js:199 (3cb0807c7510 @ 22)
Nov 21 22:57:10 fedx gnome-shell[1545]: == Stack trace for context 0x556f884ff200 ==
Nov 21 22:57:10 fedx gnome-shell[1545]: Object Clutter.Actor (0x556f8a65af90), has been already disposed — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
Nov 21 22:57:10 fedx systemd[1345]: dbus-:1.2-org.gnome.Shell.Extensions@10.service: Consumed 1.523s CPU time.
Nov 21 22:57:02 fedx gnome-shell[1545]: Could not create transient scope for PID 6811: GDBus.Error:org.freedesktop.DBus.Error.UnixProcessIdUnknown: Process with ID 6811 does not exist.
Nov 21 22:57:02 fedx gnome-shell[1545]: Window manager warning: Buggy client sent a _NET_ACTIVE_WINDOW message with a timestamp of 0 for 0x1e00004
Nov 21 22:57:01 fedx systemd[1345]: Started dbus-:1.2-org.gnome.Extensions@9.service.
Nov 21 22:57:00 fedx systemd[1345]: dbus-:1.2-org.gnome.Extensions@8.service: Consumed 2.351s CPU time.
Nov 21 22:56:58 fedx gnome-shell[1545]: #1   556f89888818 i   /home/timmy/.local/share/gnome-shell/extensions/compiz-alike-windows-effect@hermes83.github.com/extension.js:82 (1bb5254a80b0 @ 46)
Nov 21 22:56:58 fedx gnome-shell[1545]: #0   556f898888a8 i   /home/timmy/.local/share/gnome-shell/extensions/compiz-alike-windows-effect@hermes83.github.com/commonUtils.js:122 (1bb5254a8380 @ 34)
Nov 21 22:56:58 fedx gnome-shell[1545]: == Stack trace for context 0x556f884ff200 ==
Nov 21 22:56:58 fedx gnome-shell[1545]: Object .MetaWindowActorX11 (0x556f8a174330), has been already disposed — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
Nov 21 22:56:56 fedx gjs[6686]: GtkWindowControls 0x55d112dd8ac0 (windowcontrols) reported min width -166, but sizes must be >= 0
Nov 21 22:56:56 fedx gnome-shell[1545]: Can't update stage views actor <unnamed>[<MetaSurfaceActorX11>:0x556f8a179bc0] is on because it needs an allocation.
Nov 21 22:56:56 fedx gnome-shell[1545]: Can't update stage views actor <unnamed>[<MetaWindowActorX11>:0x556f8a174330] is on because it needs an allocation.
Nov 21 22:56:56 fedx gnome-shell[1545]: Can't update stage views actor <unnamed>[<MetaWindowGroup>:0x556f889c4350] is on because it needs an allocation.
Nov 21 22:56:56 fedx gjs[6686]: GtkWindowControls 0x55d112dd8ac0 (windowcontrols) reported min width -166, but sizes must be >= 0
Nov 21 22:56:49 fedx gnome-shell[1545]: Could not create transient scope for PID 6759: GDBus.Error:org.freedesktop.DBus.Error.UnixProcessIdUnknown: Process with ID 6759 does not exist.
Nov 21 22:56:49 fedx gjs[6661]: g_task_return_error: assertion '!task->ever_returned' failed
```

**Reply by Leleat** [reply]
> PS. I can no longer reproduce the crash.

Nvm, got a second freeze now. I found more errors from `compiz-alike-windows-effect` but they didn't cause any freezes when they appeared in the logs. So guess it's not caused by that extension.

The second freeze happened after I toggled the `Experimental Hover Mode` and tried to open the pie menu again. But I can't reproduce this again...?

**Comment by Schneegans** [comment]
Thank you for the detailed feedback! Let's fix all of these bugs :)

> the Super+Whatever shortcuts don't seem to work anymore

I just pushed a fix for this one. Should be working again. It was a regression as I tried to work-around another issue: If you use the <kbd>Super</kbd>+<kbd>RMB</kbd> in _Turbo Mode_ (that is: hold down <kbd>Super</kbd>, press <kbd>RMB</kbd>, move the mouse to an item, release <kbd>RMB</kbd>, release <kbd>Super</kbd>), you will toggle the overview as you released <kbd>Super</kbd> outside of Fly-Pie. Maybe a dirty fix is required which virtually releases the key if I notice that it is still held down when an action is executed...

> releasing the initial RMB in the center doesn't abort the pie menu and a quick second RMB doesn't get registered. So it seems like there is no quick way to abort/escape the pie menu. This feels especially slow, if you use the Display Timeout setting since you need to 'wait' for the pie menu to escape it.

The first issue is intentional, if we canceled the menu as soon as the initial <kbd>RMB</kbd> gets released, there would be no way of using the menu by point-and-click. The second issue seems to be that <kbd>RMB</kbd> is not registered _if <kbd>Super</kbd> is still pressed_. I guess that this can be fixed though!

> I don't know, if you have a specific reason for having the version nr in metadata.json be higher than the EGO version. 

The problem was (at least about a year ago), that the extensions app seems to re-download extensions from time to time. This would then override your git-based version. To prevent this, I am using a higher version number. But I could check whether that still happens...

> I've got a freeze/crash.

Thanks for the report! I'll hopefully be able to reproduce this. If not, we might have to debug this on your end!

**Comment by Leleat** [comment]
>> I don't know, if you have a specific reason for having the version nr in metadata.json be higher than the EGO version.
>
> The problem was (at least about a year ago), that the extensions app seems to re-download extensions from time to time. This would then override your git-based version. To prevent this, I am using a higher version number. But I could check whether that still happens...

Nvm then, I've found a recent GNOME Shell gitlab issue, where someone reported a similiar problem (accidental downgrade) with a git repo. So I guess the issue still exists. https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/4807

**Comment by Schneegans** [comment]
> Maybe a dirty fix is required which virtually releases the key if I notice that it is still held down when an action is executed...

This does not work, the "real" button release still toggles the overview. I now reverted to the Fly-Pie 9 behavior: The menu will not close as long as a modifier button is held down.

> The second issue seems to be that <kbd>RMB</kbd> is not registered if <kbd>Super</kbd> is still pressed.

This should be fixed now.

I haven't been able to reproduced the freeze you mentioned. Maybe it got fixed by waiting until the modifiers are all released?

**Reply by Leleat** [reply]
> I haven't been able to reproduced the freeze you mentioned. Maybe it got fixed by waiting until the modifiers are all released?

Didn't get any freezes since pulling this morning (after writing my comment 1h ago). Yesterday I got the 2 freezes within minutes of playing around with Fly-Pie. So (possibly), it's no longer a problem.

**Comment by Schneegans** [comment]
If tried to follow the code flow in Mutter and I think that we cannot intercept <kbd>Super</kbd>+<kbd>LMB</kbd>... Have you tried this before?

**Reply by Leleat** [reply]
> If tried to follow the code flow in Mutter and I think that we cannot intercept Super+LMB... Have you tried this before?

I believe, I did look for that code in GNOME Shell in the past but didn't find it... I guess it's completely handled in Mutter and probably can't be intercepted from an extension. But I am not a professional or experienced developer, so my opnion isn't worth much. Maybe try asking in a IRC/Matrix channel. You might be lucky and catch a GNOME dev or someone else who is knowledgable.

> This does not work, the "real" button release still toggles the overview. I now reverted to the Fly-Pie 9 behavior: The menu will not close as long as a modifier button is held down.

After trying the 2 versions for a few hours each, I definitely prefer the current behaviour over the one from this morning (where it was possible to accidently trigger the overview). It's more predictable. So everything works great for me. Thanks for the work 😀

**Reply by Schneegans** [reply]
>  Maybe try asking in a IRC/Matrix channel. You might be lucky and catch a GNOME dev or someone else who is knowledgable.

I already asked yesterday and got directed to the right location in the source. It basically starts here: https://gitlab.gnome.org/GNOME/mutter/-/blob/main/src/core/events.c#L343

> So everything works great for me. Thanks for the work :grinning:

I'll test some more as well. I'll also test for potential regressions on touch screens in the coming days. As soon as I feel confident, I'll merge this! Thanks again for the suggestion!

**Comment by Leleat** [comment]
I've noticed that rmb-ing (without `Super`) a window's titlebar also opens the pie menu instead of the old window menu.

[Edit]

Looks like you just forgot the braces around the `&` at https://github.com/Schneegans/Fly-Pie/blob/feature/super-rmb/src/extension/Shortcuts.js#L66

**Reply by Schneegans** [reply]
Good catch! Should be fixed now. Always weird that `==` has a higher precedence than `&`.

**Comment by Schneegans** [comment]
Touch screens seem to be working nicely. I'll merge this now, feel free to open a new issue if you encounter any new bug!

**Reply by Leleat** [reply]
Nice, thanks for the work!

**Comment by oredaze** [comment]
A problem with Super + RMB is that it is very common in the linux world that Super + LMB moves the window and Super + RMB resizes it, even if the second is not the gnome default. As such I have set it like that and it prevents fly-pie from activating on windows (on the desktop it still works).
Furthermore activating it with 1 keyboard key + 1 mouse key is not much different than activating it with 2 keyboard keys in terms of speed/productivity. I was about to make a topic asking for an option to activate with a single mouse button (side buttons) until I saw this topic. It is common for mice to have 2-3 side buttons that are mostly unused by people. Also this is the fastest/most productive way. I am sure it is worth spending time searching for a way to make it work. Even if it's only available under Xorg!!

**Comment by Leleat** [comment]
> A problem with Super + RMB is that it is very common in the linux world that Super + LMB moves the window and Super + RMB resizes it, even if the second is not the gnome default. As such I have set it like that and it prevents fly-pie from activating on windows (on the desktop it still works).

Problem is that it's probably not possible to change `Super`+`MMB` (nor `LMB`) from a GNOME Shell extension since that is directly handled in Mutter (I think).

> Furthermore activating it with 1 keyboard key + 1 mouse key is not much different than activating it with 2 keyboard keys in terms of speed/productivity. I was about to make a topic asking for an option to activate with a single mouse button (side buttons) until I saw this topic. It is common for mice to have 2-3 side buttons that are mostly unused by people.

IMO, it is actually better because I don't need to bind Fly-Pie to a dedicated side button. I only need to bind `Super` to a button and I can directly open the overview *and* Fly Pie with just **1 side-button** and don't need to spend an extra side button just on Fly Pie.

If you do want to bind Fly Pie to a dedicated mouse side button, you can assign the side button to a keyboard shortcut and have Fly Pie be opened using that keyboard shortcut... (Not sure how common it is for side buttons to be able to simulate key combos though)

**Reply by oredaze** [reply]
> If you do want to bind Fly Pie to a dedicated mouse side button, you can assign the side button to a keyboard shortcut and have Fly Pie be opened using that keyboard shortcut... (Not sure how common it is for side buttons to be able to simulate key combos though)

Quick search didn't produce easy way for me to do that on wayland. I am just getting converted to GNOME because of fly-pie, maybe I need to search more or go to Xorg.

**Comment by Schneegans** [comment]
> I am just getting converted to GNOME because of fly-pie, maybe I need to search more or go to Xorg.

Wow, that's a great compliment :)

Sadly, the short answer to you suggestion is that binding something to mouse buttons [is impossible in GNOME Shell](https://discourse.gnome.org/t/how-to-bind-modifier-mousebutton-in-gjs/3743/2). As @Leleat mentioned, the <kbd>Super</kbd>+<kbd>RMB</kbd> combination is an exception as this is intercepted by Mutter and extensions can override the default behavior (which opens the window menu). Other than that, there is no concept of window manager shortcuts involving mouse buttons on Mutter.

Under X11, things are a bit different, though. Here, client applications can intercept any event (which is great for coding key loggers :wink:). Therefore, binding menus to mouse buttons would be possible! However, I do not want to implement this directly in Fly-Pie, as it can be achieved quite easily with a third-party tool: [xbindkeys](http://www.nongnu.org/xbindkeys/xbindkeys.html#utilisation). Here's a [tutorial](https://www.linuxuprising.com/2019/11/how-to-bind-mouse-buttons-to-keyboard.html).

Btw, another great X11-only tool is [easystroke](https://github.com/thjaeger/easystroke/wiki).

I hope this can help you!

**Reply by oredaze** [reply]
It sure does, thanks! I would figure the button out eventually, it's linux :) But, it's best if we have less struggles heh.
Once you try radial menus and make them activate with a single side mouse button you can never go back.
